internal/build/build.go

```diff
@@ -0,0 +1,224 @@
+		wd, err := os.Getwd()
+		if err != nil {
+			return nil, errors.Wrap(err, "could not get working directory")
+		}
+
+		req.workingDir = wd
+		// build output default to stdout
+		req.out = os.Stdout
+
+		// source code default directory is current dir
+		req.sourcePaths = []string{"."}
+
+		for _, opt := range opts {
+			opt(req)
+		}
+
+		// get the application CRD containing platform and other stuff we need to build
+		var app ketchv1.App
+		if err := k8sClient.Get(ctx, types.NamespacedName{Name: req.AppName}, &app); err != nil {
+			return nil, errors.Wrap(err, "could not get app")
+		}
+
+		// get the image for the platform we will use to build
+		platformImageURI, err := getPlatformImageURI(ctx, app.Spec.Platform, k8sClient)
+		if err != nil {
+			return nil, err
+		}
+
+		buildCtx, err := newBuildContext()
+		if err != nil {
+			return nil, err
+		}
+		defer buildCtx.close()
+
+		// prepare the build directory with an archive containing sources and a docker file
+		if err := buildCtx.prepare(platformImageURI, req.workingDir, req.sourcePaths); err != nil {
+			return nil, err
+		}
+
+		// create an image that contains our built source and push image to target registry
+		resp, err := dockerCli.Build(
+			ctx,
+			&docker.BuildRequest{
+				Image:          req.Image,
+				BuildDirectory: buildCtx.BuildDir(),
+				Out:            req.out,
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		response := &CreateImageFromSourceResponse{
+			ImageURI: resp.ImageURI,
+		}
+		return response, nil
+	}
+}
+
+// Retrieve a normalized platform image URI associated with the App.
+func getPlatformImageURI(ctx context.Context, platformName string, client resourceGetter) (string, error) {
+	var platform ketchv1.Platform
+	if err := client.Get(ctx, types.NamespacedName{Name: platformName}, &platform); err != nil {
+		return "", errors.Wrap(err, "could not get platform crd %q", platformName)
+	}
+
+	return docker.NormalizeImage(platform.Spec.Image)
+}
+
+type buildContext struct {
+	ephemeralBuildDir string
+}
+
+func newBuildContext() (*buildContext, error) {
+	var ctx buildContext
+	buildDir, err := ioutil.TempDir(os.TempDir(), tempArchiveDirPrefix)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not generate temp dir for build")
+	}
+	ctx.ephemeralBuildDir = buildDir
+	return &ctx, nil
+}
+
+// Call close when done with build context to clean up file system resources used for build.
+func (bc *buildContext) close() {
+	_ = os.RemoveAll(bc.ephemeralBuildDir)
+}
+
+// BuildDir contains directory where generated Dockerfile and source archive is located
+func (bc *buildContext) BuildDir() string {
+	return bc.ephemeralBuildDir
+}
+
+// Prepare the build. Create a temp directory containing a docker file, and an archive containing source codes.
+func (bc *buildContext) prepare(platformImage string, workingDir string, sourcePaths []string) error {
+	const sourceDockerfileTemplate = `FROM {{ .PlatformImage }}
```

Comment:
I think we should avoid using hardcoded string configs.

Reply:
I disagree.  This string is small, and it is only used within the scope of the method where it is called.  I was going to use gobuffalo/packr to embed this in the binary, but I found out that go:embed will be included in Go 1.16, which is much easier to use and will be a feature of the language moving forward. So in short, leave as is until Go 1.16 drops.  I put a todo comment in the code indicating as much.